datumaro/plugins/ndr.py

```diff
@@ -0,0 +1,260 @@
+            number of outputs you want.
+            the algorithm will cut whole dataset to this amount
+            if None, return result without any modification
+        over_sample: "random" or "similarity"
+            specify the strategy when num_cut > length of the result after removal
+            if random, sample from removed data randomly
+            if similarity, select from removed data with ascending order of similarity
+        under_sample: "uniform" or "inverse"
+            specify the strategy when num_cut < length of the result after removal
+            if uniform, sample data with uniform distribution
+            if inverse, sample data with reciprocal of the number of data which have same hash key
+        Algorithm Specific for gradient
+            block_shape: tuple, (h, w)
+                for the robustness, this function will operate on blocks
+                mean and variance will be calculated on this block
+            hash_dim: int
+                dimension(or bit) of the hash function
+            sim_threshold: float
+                the threshold value for saving hash-collided samples.
+                larger value means more generous, i.e., saving more samples
+        Return
+        ---------------
+        None, other subsets combined with the result
+        """
+        super().__init__(extractor)
+        if seed:
+            np.random.seed(seed)
+        if len(extractor) < 1:
+            raise ValueError("Empty dataset")
+        if not working_subset:
+            working_subset = DEFAULT_SUBSET_NAME
+        if working_subset not in extractor.subsets():
+            raise ValueError("Invalid working_subset name")
+        self.working_subset = working_subset
+
+        # parameter validation before main runs
+        if working_subset == duplicated_subset:
+            raise ValueError("working_subset == duplicated_subset")
+        if algorithm not in ('gradient'):
+            raise ValueError("Invalid algorithm name")
+        if over_sample not in ("random", "similarity"):
+            raise ValueError("Invalid over_sample")
+        if under_sample not in ("uniform", "inverse"):
+            raise ValueError("Invalid under_sample")
+        
+        self.duplicated_subset = duplicated_subset
+        self._remove(algorithm, working_subset, num_cut, over_sample, under_sample, **kwargs)
+        
+    def _remove(self, algorithm, subset, num_cut=None, over_sample='random', under_sample='uniform',
+        **kwargs):
+        working_subset = self._extractor.get_subset(subset)
+        having_image = []
+        all_imgs = []
+        for item in working_subset:
+            if item.image.has_data:
+                having_image.append(item)
+                all_imgs.append(item.image.data)
+
+        if num_cut and num_cut > len(all_imgs):
+            raise ValueError("The number of images is smaller than the cut you want")
+        
+        if algorithm == "gradient":
```

Comment:
Consider using an enum instead of a string constant.

Reply:
Could you explain more about enum?

```python
class Algo_name(Enum):
    algo_name1 = "..."
    ....

...
if algorithm == Algo_name.algo_name1
```

Do you mean creating a class by inheriting Enum like above?

Reply:
```python
AlgoName = Enum('AlgoName', ['good', 'best', 'poor'])
```
Examples: https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/plugins/voc_format/format.py#L16, https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/cli/contexts/project/__init__.py#L228